kolena/_experimental/object_detection/evaluator.py

```diff
@@ -0,0 +1,343 @@
+    # F1-Optimal threshold computation. Subsequent requests for a given threshold strategy (for other test cases) will
+    # hit this cache and use the previously computed population level confidence thresholds.
+    threshold_cache: Dict[str, Dict[str, float]] = {}  # configuration -> label -> threshold
+
+    # Keeps track of test sample locators for each test case (used for total # of image count in aggregated metrics)
+    locators_by_test_case: Dict[str, List[str]] = {}
+
+    def compute_image_metrics(
+        self,
+        ground_truth: GroundTruth,
+        inference: Inference,
+        configuration: ThresholdConfiguration,
+        labels: Set[str],  # the labels being tested in this test case
+    ) -> TestSampleMetrics:
+        assert configuration is not None, "must specify configuration"
+        thresholds = self.get_confidence_thresholds(configuration)
+        bbox_matches: MulticlassInferenceMatches = match_inferences_multiclass(
+            ground_truth.bboxes,
+            [
+                inf
+                for inf in inference.bboxes
+                if inf.label in labels and inf.score >= max(configuration.min_confidence_score, thresholds[inf.label])
+            ],
+            ignored_ground_truths=ground_truth.ignored_bboxes,
+            mode="pascal",
+            iou_threshold=configuration.iou_threshold,
+        )
+        tp = [inf for _, inf in bbox_matches.matched if inf.label in labels]
+        fp = [inf for inf in bbox_matches.unmatched_inf if inf.label in labels]
+        fn = [gt for gt, _ in bbox_matches.unmatched_gt if gt.label in labels]
+
+        confused = [inf for gt, inf in bbox_matches.unmatched_gt if inf is not None and gt.label in labels]
+        non_ignored_inferences = [inf for _, inf in bbox_matches.matched] + bbox_matches.unmatched_inf
+        scores = [inf.score for inf in non_ignored_inferences if inf.label in labels]
+        fields = [(threshold_key(label), float) for label in thresholds.keys()]
+        dc = make_dataclass("ExtendedImageMetrics", bases=(TestSampleMetrics,), fields=fields, frozen=True)
+        return dc(
+            TP=tp,
+            FP=fp,
+            FN=fn,
+            Confused=confused,
+            count_TP=len(tp),
+            count_FP=len(fp),
+            count_FN=len(fn),
+            count_Confused=len(confused),
+            has_TP=len(tp) > 0,
+            has_FP=len(fp) > 0,
+            has_FN=len(fn) > 0,
+            has_Confused=len(confused) > 0,
+            max_confidence_above_t=max(scores) if len(scores) > 0 else None,
+            min_confidence_above_t=min(scores) if len(scores) > 0 else None,
+            **{threshold_key(label): value for label, value in thresholds.items()},
+        )
+
+    def compute_f1_optimal_thresholds(
+        self,
+        configuration: ThresholdConfiguration,
+        inferences: List[Tuple[TestSample, GroundTruth, Inference]],
+    ) -> None:
```

Comment:
This feels like it should return the optimal thresholds, not just compute them and cache them.